android-sdk/src/main/java/com/mobilecoin/lib/MobileCoinClient.java

```diff
@@ -330,6 +331,305 @@ public Amount getTransferableAmount(@NonNull TokenId tokenId) throws NetworkExce
+        }
+        Transaction spendInputTransaction = prepareTransaction(
+                accountKey.getPublicAddress(),
+                presignedInput.getPseudoOutputAmount().subtract(fee),
+                txOutToSpend,
+                fee,
+                TxOutMemoBuilder.createSenderAndDestinationRTHMemoBuilder(accountKey),
+                DefaultRng.createInstance()
+        ).getTransaction();
+        try {
+            submitTransaction(spendInputTransaction);
+        } catch(InvalidTransactionException e) {
+            switch(e.getResult()) {
+                case ContainsSpentKeyImage:
+                    Logger.i(TAG, "Failed to cancel previously spent SignedContingentInput");
+                    return SignedContingentInput.CancelationResult.FAILED_ALREADY_SPENT;
+                default:
+                    Logger.e(TAG, "Failed to cancel a SignedContingent for unknown reason");
+                    return SignedContingentInput.CancelationResult.FAILED_UNKNOWN;
+            }
+        }
+        while(getTransactionStatus(spendInputTransaction).equals(Transaction.Status.UNKNOWN)) {
+            try {
+                Thread.sleep(STATUS_CHECK_DELAY_MS);
+            } catch (InterruptedException e) {
+                Logger.e(TAG, e);
+            }
+        }
+        if(getTransactionStatus(spendInputTransaction).equals(Transaction.Status.ACCEPTED)) {
+            return SignedContingentInput.CancelationResult.SUCCESS;
+        }
+        Logger.e(TAG, "Failed to cancel a SignedContingent for unknown reason");
+        return SignedContingentInput.CancelationResult.FAILED_UNKNOWN;
+    }
+
+    @Override
+    @NonNull
+    public Transaction prepareTransaction(
+            @NonNull final SignedContingentInput presignedInput,
+            @NonNull final Amount fee
+    ) throws TransactionBuilderException, AttestationException, FogSyncException, InvalidFogResponse,
+            NetworkException, InsufficientFundsException, FragmentedAccountException, FogReportException {
+        return prepareTransaction(
+                presignedInput,
+                fee,
+                DefaultRng.createInstance()
+            );
+    }
+
+    @Override
+    @NonNull
+    public Transaction prepareTransaction(
+            @NonNull final SignedContingentInput presignedInput,
+            @NonNull final Amount fee,
+            @NonNull final Rng rng
+    ) throws TransactionBuilderException, AttestationException, FogSyncException, InvalidFogResponse,
+            NetworkException, InsufficientFundsException, FragmentedAccountException, FogReportException {
+        if(!presignedInput.isValid()) {
+            throw new TransactionBuilderException("Cannot build transaction with invalid SignedContingentInput");
+        }
+        int blockVersion = blockchainClient.getOrFetchNetworkBlockVersion();
+        if(blockVersion < 3) {
+            throw new UnsupportedOperationException("Unsupported until block version 3");
```

Comment:
It isn't easy to catch runtime exceptions, can this be a checked exception i.e. `TransactionBuilderException`?

Reply:
Yeah, good point. Let me also look to see if I have any other RuntimeExceptions that aren't appropriate.